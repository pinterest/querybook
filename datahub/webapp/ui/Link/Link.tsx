import React from 'react';
import { LocationDescriptor } from 'history';
import styled from 'styled-components';
import { Link as LinkImport, LinkProps } from 'react-router-dom';

const StyledLink = styled('a')`
    ${({ naturalLink }) =>
        naturalLink &&
        `
        color: inherit;
    `};
`;

function isInternalUrl(url: LocationDescriptor): boolean {
    if (url && typeof url === 'string' && url.length && url[0] === '/') {
        return true;
    } else if (url && typeof url === 'object') {
        return true;
    }

    return false;
}

export interface ILinkProps {
    to?: LocationDescriptor;
    onClick?: (to: React.MouseEvent) => any;
    newTab?: boolean;

    className?: string;
    naturalLink?: boolean;
    stopProgation?: boolean;

    linkProps?: Partial<LinkProps>;
}

const openNewTab = (url: string) => window.open(url);
const openInTab = (url: string) => (window.location.href = url);

export class Link extends React.PureComponent<ILinkProps> {
    public handleClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
        e.preventDefault();
        const { to, onClick, newTab } = this.props;
        const isCmdDown = e.metaKey;
        if (onClick) {
            onClick(e);
        } else if (to && typeof to === 'string') {
            if (isCmdDown || newTab) {
                openNewTab(to);
            } else {
                openInTab(to);
            }
        }
    };

    public render() {
        const {
            children,
            to,
            onClick,
            newTab,
            naturalLink,
            stopProgation,
            linkProps,
            ...elementProps
        } = this.props;

        const linkComponent = isInternalUrl(to) ? (
            <LinkImport to={to} {...elementProps} {...linkProps}>
                {children}
            </LinkImport>
        ) : (
            <StyledLink
                {...elementProps}
                href={to}
                naturalLink={naturalLink}
                onMouseDown={this.handleClick}
            >
                {children}
            </StyledLink>
        );

        if (stopProgation) {
            return (
                <span onClick={(event) => event.stopPropagation()}>
                    {linkComponent}
                </span>
            );
        }
        return linkComponent;
    }
}
