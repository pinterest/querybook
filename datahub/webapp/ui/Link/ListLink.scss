@import './../../scss_variables.scss';

.ListLink {
    display: flex;
    align-items: center;
    border-left: solid 6px transparent;
    padding: 4px 10px 4px 4px;
    color: var(--dark-text-color);
    cursor: pointer;
    user-select: none;

    .ListLinkText {
        flex-grow: 1;
        margin-left: 4px;
        margin-right: 8px;
        font-size: var(--small-text-size);
        line-height: 27px;
        @include one-line-ellipsis();
    }
    .Icon {
        flex-shrink: 1;
        svg {
            padding: 20%;
        }
    }

    &.row {
        padding: 4.5px 10px 4.5px 4.5px;
        .ListLinkText {
            line-height: normal;
        }
    }
    &.empty-title {
        .ListLinkText {
            opacity: 0.7;
            font-style: italic;
        }
    }

    &.selected {
        border-color: var(--color-accent-bg);
        font-weight: var(--bold-font);
    }
    &:hover {
        background-color: var(--color-accent-bg-hover);
        color: var(--dark-text-color);
    }
}
