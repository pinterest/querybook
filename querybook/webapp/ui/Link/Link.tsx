import { LocationDescriptor } from 'history';
import React, { useCallback } from 'react';
import { Link as LinkImport, LinkProps } from 'react-router-dom';
import styled from 'styled-components';

import { stopPropagation as stopPropagationFunc } from 'lib/utils/noop';

const StyledLink = styled('a')`
    ${({ naturalLink }) =>
        naturalLink &&
        `
        color: var(--color-accent-dark);
        &:hover {
            opacity: 0.9;
        }
        text-decoration: underline;
    `};
`;

function isInternalUrl(url: LocationDescriptor): boolean {
    if (url && typeof url === 'string' && url.length && url[0] === '/') {
        return true;
    } else if (url && typeof url === 'object') {
        return true;
    }

    return false;
}

function getExternalUrl(url: LocationDescriptor) {
    if (url && typeof url === 'string') {
        if (/^[a-zA-Z]+:\/\//.test(url)) {
            return url;
        }
        return `//${url}`;
    }

    return url;
}

// ReactRouter's Link's tying thinks that defaultValue can only be string | string[]
// This causes a ts error since defaultValue can also be a number
type AnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & {
    defaultValue?: string | string[];
};

export interface ILinkProps extends AnchorProps {
    to?: LocationDescriptor;
    onClick?: (to: React.MouseEvent) => any;
    newTab?: boolean;

    className?: string;
    naturalLink?: boolean;
    stopProgation?: boolean;
    contextmenu?: boolean;

    linkProps?: Partial<LinkProps>;
}

const openNewTab = (url: string) => window.open(url);
const openInTab = (url: string) => (window.location.href = url);

export const Link: React.FC<ILinkProps> = ({
    to,
    onClick,
    newTab,
    children,

    naturalLink,
    stopProgation,
    linkProps,
    ...elementProps
}) => {
    const handleClick = useCallback(
        (e: React.MouseEvent<HTMLAnchorElement>) => {
            e.preventDefault();
            const isCmdDown = e.metaKey;

            if (onClick) {
                onClick(e);
            } else if (to && typeof to === 'string') {
                if (isCmdDown || newTab) {
                    openNewTab(to);
                } else {
                    openInTab(to);
                }
            }
        },
        [to, onClick, newTab]
    );

    const linkComponent = isInternalUrl(to) ? (
        <LinkImport to={to} {...(elementProps as LinkProps)} {...linkProps}>
            {children}
        </LinkImport>
    ) : (
        <StyledLink
            href={getExternalUrl(to)}
            naturalLink={naturalLink}
            onClick={handleClick}
            {...elementProps}
            {...(newTab
                ? {}
                : { target: '_blank', rel: 'noopener noreferrer' })}
        >
            {children}
        </StyledLink>
    );

    if (stopProgation) {
        return <span onClick={stopPropagationFunc}>{linkComponent}</span>;
    }
    return linkComponent;
};
