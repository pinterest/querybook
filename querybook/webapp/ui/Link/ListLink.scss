@import './../../scss_variables.scss';

.ListLink {
    display: flex;
    align-items: center;
    border-left: solid 6px transparent;
    padding: 4px 10px 4px 4px;
    color: var(--dark-text-color);
    cursor: pointer;
    user-select: none;

    .ListLinkText,
    .ListLinkPlaceholder {
        flex-grow: 1;
        margin-left: 4px;
        margin-right: 12px;
        font-size: var(--small-text-size);
        line-height: 27px;
        @include one-line-ellipsis();

        &:last-child {
            margin-right: 28px;
        }
    }
    .ListLinkPlaceholder {
        opacity: 0.7;
        font-style: italic;
    }

    .Icon {
        flex-shrink: 1;
        svg {
            padding: 20%;
        }
    }

    &.row {
        padding: 4.5px 10px 4.5px 4.5px;
        .ListLinkText,
        .ListLinkPlaceholder {
            line-height: normal;
            margin-left: 0px;
        }
    }
    &.selected {
        border-color: var(--color-accent-bg);
        font-weight: var(--bold-font);
    }
    &:hover {
        background-color: var(--color-accent-bg-hover);
        color: var(--dark-text-color);
    }
}
