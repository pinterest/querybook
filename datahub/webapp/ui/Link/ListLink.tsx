import * as React from 'react';
import classNames from 'classnames';

import { Icon } from 'ui/Icon/Icon';
import { Link, ILinkProps } from './Link';

import './ListLink.scss';

interface IProps extends ILinkProps {
    className?: string;
    title?: string;
    placeholder?: string;
    icon?: string;
    isRow?: boolean;
}

export const ListLink: React.FunctionComponent<IProps> = ({
    className,
    title,
    icon,
    isRow,
    placeholder = 'Untitled',
    ...listProps
}) => {
    const mergedClassName = classNames({
        ListLink: true,
        [className]: !!className,
        row: isRow,
    });
    return (
        <Link className={mergedClassName} {...listProps}>
            {title ? (
                <span className="ListLinkText">{title}</span>
            ) : (
                <span className="ListLinkPlaceholder">{placeholder}</span>
            )}
            {icon && <Icon name={icon} />}
        </Link>
    );
};
