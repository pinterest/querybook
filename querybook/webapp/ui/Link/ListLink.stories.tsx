import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { ListLink } from './ListLink';

export const _ListLink = (args) => {
    const [selectedLink, setSelectedLink] = React.useState(0);

    return (
        <div style={{ width: '160px' }}>
            <ListLink
                {...args}
                className={selectedLink === 0 ? 'selected' : ''}
                onClick={() => setSelectedLink(0)}
                title="List Link 1"
            />
            <ListLink
                {...args}
                className={selectedLink === 1 ? 'selected' : ''}
                onClick={() => setSelectedLink(1)}
                title="List Link 2"
            />
            <ListLink
                {...args}
                className={selectedLink === 2 ? 'selected' : ''}
                onClick={() => setSelectedLink(2)}
                title="List Link 3"
            />
            <ListLink
                {...args}
                className={selectedLink === 3 ? 'selected' : ''}
                onClick={() => setSelectedLink(3)}
            />
        </div>
    );
};

_ListLink.args = {
    icon: 'zap',
    isRow: false,
};

export default {
    title: 'Stateless/ListLink',
    decorators: [centered],
};
