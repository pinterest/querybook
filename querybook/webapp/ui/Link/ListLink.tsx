import * as React from 'react';
import clsx from 'clsx';

import { Icon } from 'ui/Icon/Icon';
import { Link, ILinkProps } from './Link';

import './ListLink.scss';

interface IProps extends ILinkProps {
    className?: string;
    title?: string;
    placeholder?: string;
    icon?: string;
    isRow?: boolean;
}

export const ListLink: React.FunctionComponent<IProps> = React.memo(
    ({
        className,
        title,
        icon,
        isRow,
        placeholder = 'Untitled',
        children,
        ...listProps
    }) => {
        const mergedClassName = clsx({
            ListLink: true,
            [className]: !!className,
            row: isRow,
        });
        return (
            <Link className={mergedClassName} {...listProps}>
                {title ? (
                    <span className="ListLinkText">{title}</span>
                ) : placeholder ? (
                    <span className="ListLinkPlaceholder">{placeholder}</span>
                ) : null}
                {icon && <Icon name={icon} />}
                {children}
            </Link>
        );
    }
);
