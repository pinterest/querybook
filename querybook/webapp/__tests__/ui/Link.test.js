import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { Link } from '../../ui/Link/Link';

it('renders without crashing', () => {
    shallow(<Link to="https://test.com" />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Link to="https://test.com" />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<Link to="https://test.com" />);
        expect(output).toMatchSnapshot();
    });
});
