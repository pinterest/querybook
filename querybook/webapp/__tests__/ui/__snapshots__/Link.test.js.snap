// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<Link__StyledLink
  href="https://test.com"
  onMouseDown={[Function]}
  rel="noopener noreferrer"
  target="_blank"
/>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<a
  className="Link__StyledLink-sc-1k0sviq-0 gVoECW"
  href="https://test.com"
  onMouseDown={[Function]}
  rel="noopener noreferrer"
  target="_blank"
/>
`;
